File: python/cst_airfoil.py
"""
CST (Class function/Shape function Transformation) parameterization for airfoils.

Based on Kulfan's CST method:
  Kulfan, B.M., "Universal Parametric Geometry Representation Method",
  Journal of Aircraft, Vol. 45, No. 1, 2008, pp. 142-158.

The airfoil surface is represented as:
  y(x) = C(x) * S(x) + x * dz_te
where:
  C(x) = x^N1 * (1-x)^N2          (class function)
  S(x) = sum( A_i * K_i * x^i * (1-x)^(n-i) )  (shape function, Bernstein polynomials)
  dz_te = trailing edge thickness
  N1, N2 = 0.5, 1.0 for round-nose airfoils
"""

import numpy as np
from math import factorial
import matplotlib.pyplot as plt


def bernstein_poly(n, i, x):
    """Compute the i-th Bernstein basis polynomial of degree n at x."""
    K = factorial(n) / (factorial(i) * factorial(n - i))
    return K * x**i * (1 - x)**(n - i)


def class_function(x, N1=0.5, N2=1.0):
    """
    CST class function for airfoils.
    N1=0.5, N2=1.0 gives a round leading edge and sharp trailing edge.
    """
    return x**N1 * (1 - x)**N2


def shape_function(x, A):
    """
    CST shape function using Bernstein polynomials.
    A: array of CST coefficients (Bernstein weights).
    """
    n = len(A) - 1
    S = np.zeros_like(x)
    for i, a in enumerate(A):
        S += a * bernstein_poly(n, i, x)
    return S


def cst_airfoil(x, A_upper, A_lower, dz_te=0.0, N1=0.5, N2=1.0):
    """
    Generate airfoil coordinates using CST parameterization.

    Parameters
    ----------
    x : ndarray
        Normalized chordwise coordinates, 0 <= x <= 1.
    A_upper : array-like
        CST coefficients for the upper surface.
    A_lower : array-like
        CST coefficients for the lower surface.
    dz_te : float
        Half trailing edge thickness (applied symmetrically).
    N1, N2 : float
        Class function exponents (default 0.5, 1.0 for airfoils).

    Returns
    -------
    y_upper : ndarray
        Upper surface y-coordinates.
    y_lower : ndarray
        Lower surface y-coordinates.
    """
    A_upper = np.asarray(A_upper, dtype=float)
    A_lower = np.asarray(A_lower, dtype=float)

    C = class_function(x, N1, N2)
    S_upper = shape_function(x, A_upper)
    S_lower = shape_function(x, A_lower)

    y_upper = C * S_upper + x * dz_te
    y_lower = C * S_lower - x * dz_te

    return y_upper, y_lower


def fit_cst_airfoil(x_data, y_upper_data, y_lower_data, order=6,
                    dz_te=0.0, N1=0.5, N2=1.0):
    """
    Fit CST coefficients to existing airfoil coordinate data using least squares.

    Parameters
    ----------
    x_data : ndarray
        Chordwise coordinates (0 to 1).
    y_upper_data : ndarray
        Upper surface y-coordinates.
    y_lower_data : ndarray
        Lower surface y-coordinates.
    order : int
        Order of the Bernstein polynomial (number of coefficients = order + 1).
    dz_te : float
        Half trailing edge thickness.
    N1, N2 : float
        Class function exponents.

    Returns
    -------
    A_upper : ndarray
        Fitted upper surface CST coefficients.
    A_lower : ndarray
        Fitted lower surface CST coefficients.
    """
    C = class_function(x_data, N1, N2)
    n = order

    # Build Bernstein basis matrix
    B = np.zeros((len(x_data), n + 1))
    for i in range(n + 1):
        B[:, i] = bernstein_poly(n, i, x_data)

    # Upper surface: y_upper - x*dz_te = C * S_upper = C * B @ A_upper
    rhs_upper = y_upper_data - x_data * dz_te
    W_upper = np.diag(C) @ B
    A_upper, _, _, _ = np.linalg.lstsq(W_upper, rhs_upper, rcond=None)

    # Lower surface: y_lower + x*dz_te = C * S_lower = C * B @ A_lower
    rhs_lower = y_lower_data + x_data * dz_te
    W_lower = np.diag(C) @ B
    A_lower, _, _, _ = np.linalg.lstsq(W_lower, rhs_lower, rcond=None)

    return A_upper, A_lower


def plot_airfoil(x, y_upper, y_lower, title="CST Airfoil"):
    """Plot the airfoil shape."""
    fig, ax = plt.subplots(1, 1, figsize=(10, 4))
    ax.plot(x, y_upper, 'b-', linewidth=1.5, label='Upper surface')
    ax.plot(x, y_lower, 'r-', linewidth=1.5, label='Lower surface')
    ax.set_xlabel('x/c')
    ax.set_ylabel('y/c')
    ax.set_title(title)
    ax.set_aspect('equal')
    ax.legend()
    ax.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.show()


# ---- Example usage ----
if __name__ == "__main__":
    # Define CST coefficients (example: NACA 0012-like)
    A_upper = [0.17, 0.15, 0.16, 0.14, 0.12, 0.10, 0.08]
    A_lower = [-0.17, -0.15, -0.16, -0.14, -0.12, -0.10, -0.08]
    dz_te = 0.001  # small trailing edge gap

    # Generate airfoil coordinates
    x = np.linspace(0, 1, 201)
    # Use cosine spacing for better LE resolution
    theta = np.linspace(0, np.pi, 201)
    x_cos = 0.5 * (1 - np.cos(theta))

    y_upper, y_lower = cst_airfoil(x_cos, A_upper, A_lower, dz_te)

    print("=== CST Airfoil Parameterization ===")
    print(f"Upper surface coefficients: {A_upper}")
    print(f"Lower surface coefficients: {A_lower}")
    print(f"Trailing edge gap: {2 * dz_te}")
    print(f"Number of points: {len(x_cos)}")
    print(f"Max thickness: {np.max(y_upper - y_lower):.6f}")

    # Plot the airfoil
    plot_airfoil(x_cos, y_upper, y_lower, title="CST Parameterized Airfoil")

    # Demonstrate fitting: fit CST to the generated airfoil
    A_upper_fit, A_lower_fit = fit_cst_airfoil(x_cos, y_upper, y_lower, order=6, dz_te=dz_te)
    print(f"\nFitted upper coefficients: {np.round(A_upper_fit, 6)}")
    print(f"Fitted lower coefficients: {np.round(A_lower_fit, 6)}")

    y_upper_fit, y_lower_fit = cst_airfoil(x_cos, A_upper_fit, A_lower_fit, dz_te)
    error = np.max(np.abs(y_upper - y_upper_fit)) + np.max(np.abs(y_lower - y_lower_fit))
    print(f"Max fitting error: {error:.2e}")

File: matlab/cst_airfoil.m
%% CST Airfoil Parameterization
% CST (Class function/Shape function Transformation) method for airfoils.
%
% Based on Kulfan's CST method:
%   Kulfan, B.M., "Universal Parametric Geometry Representation Method",
%   Journal of Aircraft, Vol. 45, No. 1, 2008, pp. 142-158.
%
% The airfoil surface is represented as:
%   y(x) = C(x) * S(x) + x * dz_te
% where:
%   C(x) = x^N1 * (1-x)^N2          (class function)
%   S(x) = sum( A_i * K_i * x^i * (1-x)^(n-i) )  (shape function)

function cst_airfoil()
    %% Main script
    clc; close all;

    % CST coefficients (example: NACA 0012-like)
    A_upper = [0.17, 0.15, 0.16, 0.14, 0.12, 0.10, 0.08];
    A_lower = [-0.17, -0.15, -0.16, -0.14, -0.12, -0.10, -0.08];
    dz_te = 0.001;  % half trailing edge gap

    % Cosine-spaced x coordinates for better LE resolution
    n_pts = 201;
    theta = linspace(0, pi, n_pts);
    x = 0.5 * (1 - cos(theta));

    % Generate airfoil
    [y_upper, y_lower] = compute_cst_airfoil(x, A_upper, A_lower, dz_te);

    % Display info
    fprintf('=== CST Airfoil Parameterization ===\n');
    fprintf('Upper surface coefficients: [');
    fprintf('%.2f ', A_upper);
    fprintf(']\n');
    fprintf('Lower surface coefficients: [');
    fprintf('%.2f ', A_lower);
    fprintf(']\n');
    fprintf('Trailing edge gap: %.4f\n', 2 * dz_te);
    fprintf('Number of points: %d\n', n_pts);
    fprintf('Max thickness: %.6f\n', max(y_upper - y_lower));

    % Plot airfoil
    plot_airfoil(x, y_upper, y_lower, 'CST Parameterized Airfoil');

    % Demonstrate fitting
    order = 6;
    [A_upper_fit, A_lower_fit] = fit_cst(x, y_upper, y_lower, order, dz_te);
    fprintf('\nFitted upper coefficients: [');
    fprintf('%.6f ', A_upper_fit);
    fprintf(']\n');
    fprintf('Fitted lower coefficients: [');
    fprintf('%.6f ', A_lower_fit);
    fprintf(']\n');

    [y_upper_fit, y_lower_fit] = compute_cst_airfoil(x, A_upper_fit, A_lower_fit, dz_te);
    err = max(abs(y_upper - y_upper_fit)) + max(abs(y_lower - y_lower_fit));
    fprintf('Max fitting error: %.2e\n', err);
end

%% Class function
function C = class_func(x, N1, N2)
    if nargin < 2, N1 = 0.5; end
    if nargin < 3, N2 = 1.0; end
    C = x.^N1 .* (1 - x).^N2;
end

%% Shape function using Bernstein polynomials
function S = shape_func(x, A)
    n = length(A) - 1;
    S = zeros(size(x));
    for i = 0:n
        K = nchoosek(n, i);
        B = K .* x.^i .* (1 - x).^(n - i);
        S = S + A(i + 1) * B;
    end
end

%% Compute CST airfoil coordinates
function [y_upper, y_lower] = compute_cst_airfoil(x, A_upper, A_lower, dz_te, N1, N2)
    if nargin < 4, dz_te = 0.0; end
    if nargin < 5, N1 = 0.5; end
    if nargin < 6, N2 = 1.0; end

    C = class_func(x, N1, N2);
    S_upper = shape_func(x, A_upper);
    S_lower = shape_func(x, A_lower);

    y_upper = C .* S_upper + x * dz_te;
    y_lower = C .* S_lower - x * dz_te;
end

%% Fit CST coefficients to airfoil data
function [A_upper, A_lower] = fit_cst(x, y_upper, y_lower, order, dz_te, N1, N2)
    if nargin < 5, dz_te = 0.0; end
    if nargin < 6, N1 = 0.5; end
    if nargin < 7, N2 = 1.0; end

    n = order;
    C = class_func(x, N1, N2);

    % Build Bernstein basis matrix
    B = zeros(length(x), n + 1);
    for i = 0:n
        K = nchoosek(n, i);
        B(:, i + 1) = K .* x(:).^i .* (1 - x(:)).^(n - i);
    end

    % Weighted matrix
    W = diag(C(:)) * B;

    % Upper surface
    rhs_upper = (y_upper(:) - x(:) * dz_te);
    A_upper = (W \ rhs_upper)';

    % Lower surface
    rhs_lower = (y_lower(:) + x(:) * dz_te);
    A_lower = (W \ rhs_lower)';
end

%% Plot airfoil
function plot_airfoil(x, y_upper, y_lower, fig_title)
    figure('Position', [100 100 800 400]);
    hold on;
    plot(x, y_upper, 'b-', 'LineWidth', 1.5, 'DisplayName', 'Upper surface');
    plot(x, y_lower, 'r-', 'LineWidth', 1.5, 'DisplayName', 'Lower surface');
    xlabel('x/c');
    ylabel('y/c');
    title(fig_title);
    axis equal;
    legend('Location', 'best');
    grid on;
    hold off;
end
